fs/open_options.go

```diff
@@ -230,6 +236,150 @@ func (o *HashesOption) Mandatory() bool {
 	return false
 }
 
+// OptionResume defines a Put/Upload for doing resumes
+type OptionResume struct {
+	ID           string // resume this ID if set
+	Pos          int64  // and resume from this position
+	Src          Object
+	F            Fs
+	Remote       string
+	CacheCleaned bool
+	CacheDir     string
+}
+
+// SetID will be called by backend's Put/Update function if the object's upload
+// could be resumed upon failure
+//
+// SetID takes the passed resume ID, hash state, hash name and Fingerprint of the object and stores it in
+// --cache-dir so that future Copy operations can resume the upload if it fails
+func (o *OptionResume) SetID(ctx context.Context, ID, hashName, hashState string) error {
+	ci := GetConfig(ctx)
+	rootCacheDir := filepath.Join(o.CacheDir, "resume")
+	// Get the Fingerprint of the src object so that future Copy operations can ensure the
+	// object hasn't changed before resuming an upload
+	fingerprint := Fingerprint(ctx, o.Src, true)
+	data, err := marshalResumeJSON(ctx, fingerprint, ID, hashName, hashState)
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal data JSON")
+	}
+	if len(data) < int(ci.MaxResumeCacheSize) {
+		// Each remote will have its own directory for cached resume files
```

Comment:
We should probably factor the code out of the VFS layer which makes a filename for a cache directory based on an Fs. Its needed fixing sooo many times so re-implementing it doesn't seem like a good idea!